src/safeds_runner/server/memoization_map.py

```diff
@@ -0,0 +1,327 @@
+
+        self._update_stats_on_miss(
+            function_name,
+            access_timestamp,
+            lookup_time,
+            computation_time,
+            memory_size,
+        )
+
+        logging.info(
+            "New memoization stats for %s: (access_timestamp=%s, lookup_time=%s, computation_time=%s, memory_size=%s)",
+            key[0],
+            access_timestamp,
+            lookup_time,
+            computation_time,
+            memory_size,
+        )
+
+        return computed_value
+
+    def _create_memoization_key(
+        self,
+        function_name: str,
+        parameters: list[Any],
+        hidden_parameters: list[Any],
+    ) -> MemoizationKey:
+        """
+        Convert values provided to a memoized function call to a memoization key.
+
+        Parameters
+        ----------
+        function_name
+            Fully qualified function name
+        parameters
+            List of parameters passed to the function
+        hidden_parameters
+            List of parameters not passed to the function
+
+        Returns
+        -------
+        A memoization key, which contains the lists converted to tuples
+        """
+        return function_name, _convert_list_to_tuple(parameters), _convert_list_to_tuple(hidden_parameters)
+
+    def _lookup_value(self, key: MemoizationKey) -> Any | None:
+        """
+        Lookup a potentially existing value from the memoization cache.
+
+        Parameters
+        ----------
+        key
+            Memoization Key
+
+        Returns
+        -------
+        The value corresponding to the provided memoization key, if any exists.
+        """
+        return self._map_values.get(key)
+
+    def _compute_and_memoize_value(
+        self,
+        key: MemoizationKey,
+        function_callable: Callable,
+        parameters: list[Any],
+    ) -> Any:
+        """
+        Memoize a new function call and return computed the result.
+
+        Parameters
+        ----------
+        key
+            Memoization Key
+        function_callable
+            Function that will be called
+        parameters
+            List of parameters passed to the function
+
+        Returns
+        -------
+        The newly computed value corresponding to the provided memoization key
+        """
+        result = function_callable(*parameters)
+        self._map_values[key] = result
+        return result
+
+    def _update_stats_on_hit(self, function_name: str, access_timestamp: int, lookup_time: int) -> None:
+        """
+        Update the memoization stats on a cache hit.
+
+        Parameters
+        ----------
+        function_name
+            Fully qualified function name
+        access_timestamp
+            Timestamp when this value was last accessed
+        lookup_time
+            Duration the comparison took in nanoseconds
+        """
+        stats = self._map_stats[function_name]
+        stats.update_on_hit(access_timestamp, lookup_time)
```

Comment:
```suggestion
        stats.update_on_hit(access_timestamp, lookup_time)
        self._map_stats[function_name] = stats
```

I think this call is still needed, as otherwise the update will not by synced across processes

Reply:
Interesting, but isn't the object we assign to the dict the same object we've taken out of the dict?

Reply:
That's true, but the multiprocessing manager (dict) can only detect changes that are directly done to the dict itself.
> If standard (non-proxy) [list](https://docs.python.org/3.11/library/stdtypes.html#list) or [dict](https://docs.python.org/3.11/library/stdtypes.html#dict) objects are contained in a referent, modifications to those mutable values will not be propagated through the manager because the proxy has no way of knowing when the values contained within are modified

Source: https://docs.python.org/3.11/library/multiprocessing.html#multiprocessing.managers.SyncManager.list